katz/src/main/kotlin/katz/data/Kleisli.kt

```diff
@@ -1,30 +1,36 @@
 package katz
 
-class Kleisli<F, D, A>(val run: (D) -> HK<F, A>) {
+typealias ReaderT<F, D, A> = Kleisli<F, D, A>
 
-    companion object Factory {
+class Kleisli<F, D, A>(val MF : Monad<F>, val run: (D) -> HK<F, A>) {
 
-        fun <F, D, A> pure(x: A, f: Applicative<F>): Kleisli<F, D, A> = Kleisli { _ -> f.pure(x) }
+    fun <B> map(f: (A) -> B): Kleisli<F, D, B> = Kleisli(MF, { a -> MF.map(run(a), f) })
 
-        fun <F, D> ask(f: Applicative<F>): Kleisli<F, D, D> = Kleisli { f.pure(it) }
-    }
+    fun <B> flatMap(f: (A) -> Kleisli<F, D, B>): Kleisli<F, D, B> =
+            Kleisli(MF, { d ->
+                MF.flatMap(run(d)) { a -> f(a).run(d) }
+            })
+
+    inline fun <reified F, D, A> Kleisli<F, D, Kleisli<F, D, A>>.flatten(): Kleisli<F, D, A> =
+            flatMap({ it })
+
+    fun <B> zip(o: Kleisli<F, D, B>): Kleisli<F, D, Pair<A, B>> =
+            flatMap({ a ->
+                o.map({ b -> Pair(a, b) })
+            })
+
+    fun <DD> local(f: (DD) -> D): Kleisli<F, DD, A> = Kleisli(MF, { dd -> run(f(dd)) })
 
-    fun <B> map(ft: Functor<F>, f: (A) -> B): Kleisli<F, D, B> = Kleisli { a -> ft.map(run(a), f) }
+    companion object {
 
-    fun <B> flatMap(m: Monad<F>, f: (A) -> Kleisli<F, D, B>): Kleisli<F, D, B> =
-            Kleisli { d ->
-                m.flatMap(run(d)) { a -> f(a).run(d) }
-            }
+        inline operator fun <reified F, D, A> invoke(MF : Monad<F> = monad<F>(), noinline run: (D) -> HK<F, A>): Kleisli<F, D, A> = Kleisli(MF, run)
```

Comment:
![](https://img.memesuper.com/7a1c70861550a11ba91d51dc7b352df9_sir-the-magic-guy-gif-meme-is-meme-magic-gif_320-179.gif)